### 3.6 - Projeto Prático

In [2]:
import numpy as np
import pandas as pd

In [3]:
# Importando a tabela do Apêndice I
# A tabela contém os dados de treinamento do Perceptron
dataframe = pd.read_csv('../data/tabela_apendice_i.csv', sep=',')
dataframe.head()

x1      x2       x3    d
0 -0.6508  0.1097   4.0009 -1.0
1 -1.4492  0.8896   4.4005 -1.0
2  2.0850  0.6876  12.0710 -1.0
3  0.2626  1.1476   7.7985  1.0
4  0.6418  1.0234   7.0427  1.0

In [4]:
# Função de ativação do Perceptron
def activation_func(u: float) -> int:
    return 1 if u >= 0 else -1

In [20]:
w_all = []
# Para manter os dados consistentes, criei uma seed para sempre ter os mesmos pesos quando executado.
rng = np.random.default_rng(42)

for row in dataframe.iterrows():
    w = rng.random(4)
    x1, x2, x3, d = row[1]
    x = np.array([-1, x1, x2, x3])  # Incluindo o bias como o primeiro elemento
    u = (w * x).sum()
    y = activation_func(u)
    epochs = 1
    print(f"Pesos Iniciais: {w}")
    while y != d:
        w = w + 0.01 * (d - y) * x
        u = (w * x).sum()
        y = activation_func(u)
        epochs += 1

    w_all.append((w, epochs))

for w, e in w_all:
    print(f"Pesos Finais: {w}, Épocas: {e}")


Pesos Iniciais: [0.77395605 0.43887844 0.85859792 0.69736803]
Pesos Iniciais: [0.09417735 0.97562235 0.7611397  0.78606431]
Pesos Iniciais: [0.12811363 0.45038594 0.37079802 0.92676499]
Pesos Iniciais: [0.64386512 0.82276161 0.4434142  0.22723872]
Pesos Iniciais: [0.55458479 0.06381726 0.82763117 0.6316644 ]
Pesos Iniciais: [0.75808774 0.35452597 0.97069802 0.89312112]
Pesos Iniciais: [0.7783835  0.19463871 0.466721   0.04380377]
Pesos Iniciais: [0.15428949 0.68304895 0.74476216 0.96750973]
Pesos Iniciais: [0.32582536 0.37045971 0.46955581 0.18947136]
Pesos Iniciais: [0.12992151 0.47570493 0.22690935 0.66981399]
Pesos Iniciais: [0.43715192 0.8326782  0.7002651  0.31236664]
Pesos Iniciais: [0.8322598  0.80476436 0.38747838 0.2883281 ]
Pesos Iniciais: [0.6824955  0.13975248 0.1999082  0.00736227]
Pesos Iniciais: [0.78692438 0.66485086 0.70516538 0.78072903]
Pesos Iniciais: [0.45891578 0.5687412  0.139797   0.11453007]
Pesos Iniciais: [0.66840296 0.47109621 0.56523611 0.76499886]
Pesos In

### Resultado dos treinamentos do Perceptron
| Treinamento | $w_{0}$ (inicial) | $w_{1}$ (inicial) | $w_{2}$ (inicial) | $w_{3}$ (inicial) | $w_{0}$ (final) | $w_{1}$ (final) | $w_{2}$ (final) | $w_{3}$ (final) | Épocas |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | 0.7739 | 0.4388 | 0.8585 | 0.6973 | 0.8939 | 0.5169 | 0.8454 | 0.2172 | 7 |
| 2 | 0.0941 | 0.9756 | 0.7611 | 0.7860 | 0.2141 | 1.1495 | 0.6543 | 0.2580 | 7 |
| 3 | 0.1281 | 0.4503 | 0.3707 | 0.9267 | 0.2281 | 0.2418 | 0.3020 | -0.2803 | 6 |
| 4 | 0.6438 | 0.8227 | 0.4434 | 0.2272 | 0.6438 | 0.8227 | 0.4434 | 0.2272 | 1 |
| 5 | 0.5545 | 0.0638 | 0.8276 | 0.6316 | 0.5545 | 0.0638 | 0.8276 | 0.6316 | 1 |
| 6 | 0.7580 | 0.3545 | 0.9706 | 0.8931 | 0.8780 | 0.3236 | 0.8899 | -0.1060 | 7 |
| 7 | 0.7783 | 0.1946 | 0.4667 | 0.0438 | 0.7783 | 0.1946 | 0.4667 | 0.0438 | 1 |
| 8 | 0.1542 | 0.6830 | 0.7447 | 0.9675 | 0.2942 | 0.6702 | 0.6971 | -0.0219 | 8 |
| 9 | 0.3258 | 0.3704 | 0.4695 | 0.1894 | 0.3258 | 0.3704 | 0.4695 | 0.1894 | 1 |
| 10 | 0.1299 | 0.4757 | 0.2269 | 0.6698 | 0.1299 | 0.4757 | 0.2269 | 0.6698 | 1 |
| 11 | 0.4371 | 0.8326 | 0.7002 | 0.3123 | 0.4971 | 0.8066 | 0.6590 | -0.1813 | 4 |
| 12 | 0.8322 | 0.8047 | 0.3874 | 0.2883 | 0.8322 | 0.8047 | 0.3874 | 0.2883 | 1 |
| 13 | 0.6824 | 0.1397 | 0.1999 | 0.0073 | 0.6824 | 0.1397 | 0.1999 | 0.0073 | 1 |
| 14 | 0.7869 | 0.6648 | 0.7051 | 0.7807 | 0.9069 | 0.6157 | 0.7203 | 0.1205 | 7 |
| 15 | 0.4589 | 0.5687 | 0.1397 | 0.1145 | 0.4789 | 0.5399 | 0.1365 | -0.0571 | 2 |
| 16 | 0.6684 | 0.4710 | 0.5652 | 0.7649 | 0.7484 | 0.5440 | 0.5810 | 0.5893 | 5 |
| 17 | 0.6347 | 0.5535 | 0.5592 | 0.3039 | 0.6347 | 0.5535 | 0.5592 | 0.3039 | 1 |
| 18 | 0.0308 | 0.4367 | 0.2145 | 0.4085 | 0.0308 | 0.4367 | 0.2145 | 0.4085 | 1 |
| 19 | 0.8534 | 0.2339 | 0.0583 | 0.2813 | 0.8534 | 0.2339 | 0.0583 | 0.2813 | 1 |
| 20 | 0.2935 | 0.6619 | 0.5570 | 0.7838 | 0.2935 | 0.6619 | 0.5570 | 0.7838 | 1 |
| 21 | 0.6643 | 0.4063 | 0.8140 | 0.1669 | 0.6843 | 0.4086 | 0.8095 | 0.0221 | 2 |
| 22 | 0.0227 | 0.0900 | 0.7223 | 0.4618 | 0.0227 | 0.0900 | 0.7223 | 0.4618 | 1 |
| 23 | 0.1612 | 0.5010 | 0.1523 | 0.6963 | 0.1612 | 0.5010 | 0.1523 | 0.6963 | 1 |
| 24 | 0.4461 | 0.3810 | 0.3015 | 0.6302 | 0.4461 | 0.3810 | 0.3015 | 0.6302 | 1 |
| 25 | 0.3618 | 0.0876 | 0.1180 | 0.9618 | 0.3618 | 0.0876 | 0.1180 | 0.9618 | 1 |
| 26 | 0.9085 | 0.6997 | 0.2658 | 0.9691 | 1.0685 | 0.6363 | 0.2486 | 0.0632 | 9 |
| 27 | 0.7787 | 0.7168 | 0.4493 | 0.2722 | 0.8187 | 0.7209 | 0.4254 | -0.0150 | 3 |
| 28 | 0.0963 | 0.9026 | 0.4557 | 0.2023 | 0.0963 | 0.9026 | 0.4557 | 0.2023 | 1 |
| 29 | 0.3059 | 0.5792 | 0.1767 | 0.8566 | 0.4059 | 0.3777 | 0.1148 | -0.2360 | 6 |
| 30 | 0.7585 | 0.7194 | 0.4320 | 0.6273 | 0.7585 | 0.7194 | 0.4320 | 0.6273 | 1 |